firewood/src/merkle.rs

```diff
@@ -1175,6 +1178,285 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
+
+                let (found_node, parents) = self
+                    .merkle
+                    .get_node_and_parents_by_key(root_node, &key)
+                    .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                let Some(last_node) = found_node else {
+                    return Poll::Ready(None);
+                };
+
+                let returned_key_value = match last_node.inner() {
+                    NodeType::Branch(branch) => (key, branch.value.to_owned().unwrap().to_vec()),
+                    NodeType::Leaf(leaf) => (key, leaf.1.to_vec()),
+                    NodeType::Extension(_) => todo!(),
+                };
+
+                self.key_state = IteratorState::Iterating { last_node, parents };
+
+                return Poll::Ready(Some(Ok(returned_key_value)));
+            }
+            IteratorState::Iterating {
+                last_node,
+                mut parents,
+            } => {
+                match last_node.inner() {
+                    NodeType::Branch(branch) => {
+                        // previously rendered the value from a branch node, so walk down to the first available child
+                        if let Some(child_position) =
+                            branch.children.iter().position(|&addr| addr.is_some())
+                        {
+                            let child_address = branch.children[child_position].unwrap();
+
+                            parents.push((last_node, child_position as u8)); // remember where we walked down from
+
+                            let current_node = self
+                                .merkle
+                                .get_node(child_address)
+                                .map_err(|e| api::Error::InternalError(e.into()))?;
+
+                            let found_key = key_from_parents(&parents);
+
+                            self.key_state = IteratorState::Iterating {
+                                // continue iterating from here
+                                last_node: current_node,
+                                parents,
+                            };
+
+                            found_key
+                        } else {
+                            // Branch node with no children?
+                            return Poll::Ready(Some(Err(api::Error::InternalError(Box::new(
+                                MerkleError::ParentLeafBranch,
+                            )))));
+                        }
+                    }
+                    NodeType::Leaf(leaf) => {
+                        let mut next = parents.pop();
+                        loop {
+                            match next {
+                                None => return Poll::Ready(None),
+                                Some((parent, child_position)) => {
+                                    // Assume all parents are branch nodes
+                                    let children = parent.inner().as_branch().unwrap().chd();
```

Comment:
🤔, not for this PR, but once we remove the extension, the parent array can literally just be an array of `BranchNode` instead of `NodeType`.